server/src/database/migrations/001-initial-migration.ts

```diff
@@ -0,0 +1,384 @@
+// SPDX-FileCopyrightText: 2022 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+import { QueryInterface, DataTypes, Transaction } from 'sequelize';
+
+export default {
+  up: async (queryInterface: QueryInterface): Promise<void> => {
+    const transaction: Transaction = await queryInterface.sequelize.transaction();
+    try {
+      await queryInterface.createTable('user', {
+        id: {
+          type: DataTypes.INTEGER,
+          autoIncrement: true,
+          primaryKey: true
+        },
+        student_id: {
+          type: new DataTypes.STRING,
+          unique: true,
+          allowNull: false
+        },
+        name: {
+          type: new DataTypes.STRING,
+          allowNull: false
+        },
+        email: {
+          type: new DataTypes.STRING(255),
+          allowNull: false,
+          unique: true,
+          validate: {
+            isEmail: true
+          }
+        },
+        password: {
+          type: new DataTypes.CHAR(60),
+          allowNull: false
+        },
+        created_at: DataTypes.DATE,
+        updated_at: DataTypes.DATE
+      }, { transaction });
+      await queryInterface.createTable('course', {
+        id: {
+          type: DataTypes.INTEGER,
+          autoIncrement: true,
+          primaryKey: true
+        },
+        course_code: {
+          type: new DataTypes.STRING,
+          allowNull: false
+        },
+        min_credits: {
+          type: new DataTypes.INTEGER,
+          allowNull: false
+        },
+        max_credits: {
+          type: new DataTypes.INTEGER,
+          allowNull: false,
+        },
+        created_at: DataTypes.DATE,
+        updated_at: DataTypes.DATE
+      }, { transaction });
+      await queryInterface.createTable('course_instance', {
+        id: {
+          type: DataTypes.INTEGER,
+          autoIncrement: true,
+          primaryKey: true
+        },
+        course_id: {
+          type: DataTypes.INTEGER,
+          allowNull: false,
+          references: {
+            model: 'course',
+            key: 'id'
+          },
+          onDelete: 'CASCADE',
+          onUpdate: 'CASCADE'
+        },
+        grading_type: {
+          type: DataTypes.ENUM('PASSFAIL', 'NUMERICAL'),
+          allowNull: false
+        },
+        teaching_period: {
+          type: DataTypes.STRING,
+          allowNull: false
+        },
+        teaching_method: {
+          type: DataTypes.ENUM('LECTURE', 'EXAM'),
+          allowNull: false
+        },
+        responsible_teacher: {
+          type: DataTypes.STRING,
+          allowNull: false
+        },
+        start_date: {
+          type: new DataTypes.DATEONLY,
+          allowNull: false
+        },
+        end_date: {
```

Comment:
Sequelize [documentation](https://sequelize.org/docs/v6/core-concepts/validations-and-constraints/) is quite brief on how to set the check constraint using sequelize model. Most likely it would just add the check constraint during the table creation. We could do the same with raw queries using:

```
ALTER TABLE course_instance 
ADD CONSTRAINT start_end_date_check 
CHECK (end_date >= start_date);
```

This concerns also the other models with constraint requirements.

Reply:
There exists a function `queryInterface.addConstraint`, but it seems that it's not possible to define check constraints in the Sequelize model: https://github.com/sequelize/sequelize/issues/11211

Reply:
constraints added [3eb7287189549d9cd70881f7a29b471b305f7c38](https://github.com/aalto-grades/base-repository/pull/51/commits/3eb7287189549d9cd70881f7a29b471b305f7c38)